rubicon/java/android_events.py

```diff
@@ -69,16 +86,30 @@ def _run_event_loop_once_cooperatively(self):
             asyncio.events._set_running_loop(None)
             self._set_coroutine_origin_tracking(False)
             sys.set_asyncgen_hooks(*self._old_agen_hooks)
-            # Remove Android-specific object.
-            self._handler = None
+            # Remove Android event loop interop objects.
+            self.android_interop = None
+            return
 
         # If we have actually already stopped, then do nothing.
         if self._thread_id is None:
             return
 
-        self._event_loop_will_run_soon = False
+        timeout = self._get_next_delayed_task_wakeup()
+        if timeout is None:
+            # No delayed tasks.
+            return
 
-        # Based heavily on `BaseEventLoop._run_once()` from CPython.
+        # Ask Android to wake us up to run delayed tasks. Running delayed tasks also
+        # checks for other tasks that require wakeup by calling this method. The fact that
+        # running delayed tasks can trigger the next wakeup is what makes this event loop a "loop."
+        self.android_interop.call_later(
+            "run_delayed_tasks", self.run_delayed_tasks, timeout * 1000,
+        )
+
+    def _get_next_delayed_task_wakeup(self):
+        """Compute the time to sleep before we should be woken up to handle delayed tasks."""
+        # This is based heavily on the CPython's implementation of `BaseEventLoop._run_once()`
+        # before it blocks on `select()`.
```

Comment:
I'm assuming there's a potentially upstreamable patch here that factor out the part of this method that is common to a "non-select, cooperative" event loop? Long term, if we can avoid the copypasta, we can reduce our maintenance overhead.

Reply:
Big +1